### FISH Probe Design Pipelines + SLURM-based parallelization

<br><br>

1. Standard Workflow


blabla, summarized descriptions will go here


<br><br>

---- 
### Standardard Workflow

Usage:

The pipeline consists of a `main.sh` that manages a series of modules. <br>
The following bash variables must be adjusted manually:

 | Variable | Description | TO-DO |
 | -------- | ----------- | ---- |
 | GENOME      | <br>Path to genome annotation `.fa` / `.fa.gz`<br>Index `.fai` / `.gzi` files must also exist.<br><br> | / |
 | ANNOT_INPUT | <br>Path to custom gene annotation `.gtf` / `.gtf.gz`<br><br>              | Add example / description of input ? |
 | WORKDIR | <br>Path to working / output directory<br><br>    | Add `.` as default if nothing is specified? |
 | PIPELINE_MODS | <br>Path to folder with modules<br><br> | / |


blabla
